Creates a BYO Sink Example Guide (#4208)
knative/docs: docs/developer/eventing/sinks/creating-event-sinks/writing-event-sinks-with-golang/README.md

---
title: "Writing an event sink using Golang"
weight: 10
type: "docs"
showlandingtoc: "false"
---

# Writing an event sink using Golang

This tutorial provides instructions to build an event sink in Golang and implement it as a ksvc.

## A Quick Overview of Building and Deploying a Simple Event Sink with Golang

Below we will create a simple Sink that receives CloudEvents and performs an example operation on them.

From within the `$GOPATH`, Create a folder `demosink` with a new file `main.go`:

> ```go

Author: I personally like all the code inline. but I am okay with it living next to this document if that is what the consensus is.

	code := CodeSuccess
	ceResult := cloudevents.ResultACK

	result, err := processEvent(e)
	// this error handling is for demo purposes only, since processEvent()
	// always succeeds
	if err != nil {
		code = CodeFailure
		ceResult = cloudevents.ResultNACK
	}

	return newAckEvent(e, code, result), ceResult
}

// processEvent processes the event and returns the result of the processing.
func processEvent(e cloudevents.Event) (interface{} /*result*/, error) {
	tBegin := time.Now()

	// simulate processing time
	randomDelay()

	res := &exampleResult{
		Message:        "event processed successfully",
		ProcessingTime: time.Since(tBegin).Milliseconds(),
	}

	return res, nil
}

// exampleResult represents a fictional structured result of some event
// processing.
type exampleResult struct {
	Message        string `json:"message"`
	ProcessingTime int64  `json:"processing_time_ms"`
}

// newAckEvent returns a CloudEvent that acknowledges the processing of an
// event.
func newAckEvent(e cloudevents.Event, code ACKCode, detail interface{}) *cloudevents.Event {
	resp := cloudevents.NewEvent()
	resp.SetType(eventTypeAck)
	resp.SetSource(eventSrcName)

> 	resp.SetExtension(ceExtProcessedType, e.Type())

Author: as an example. should we not?

> 	resp.SetExtension(ceExtProcessedType, e.Type())

Reviewer: KISS unless there's something special you want to highlight, and that should be documented.

	resp.SetExtension(ceExtProcessedSource, e.Source())
	resp.SetExtension(ceExtProcessedID, e.ID())

	data := &ACKResponse{
		Code:   code,
		Detail: detail,
	}

	if err := resp.SetData(cloudevents.ApplicationJSON, data); err != nil {
		log.Panicf("Error serializing CloudEvent data: %s", err)
	}

	return &resp
}

// randomDelay sleeps for a random amount of time.
func randomDelay() {
	const maxDelay = 100 * time.Millisecond

	randomDelay := time.Duration(rand.Int63n(int64(maxDelay)))
	time.Sleep(randomDelay)
}

```

Create the `go.mod` file and run `tidy`:
```
go mod init
go mod tidy
```

Create a new file `Dockerfile` in the same folder:
```dockerfile
FROM golang:1.15-buster AS builder
ENV CGO_ENABLED 0
WORKDIR /project
COPY . ./
RUN go build
FROM gcr.io/distroless/static:nonroot
COPY --from=builder /project/demosink /
ENTRYPOINT ["/demosink"]
```

Build the image:

Google:
```cmd

> gcloud builds submit --tag gcr.io/<project-name>/<image-name> .

Author: no. it will build as is.

       -H "Content-Type: application/json" \
       -d '{"fromName":"richard","toName":"bob","message":"hello"}'
```

We can represent the expected data in this event as a struct in Golang, 
add this struct to the `main.go` file.
```go
type EventData struct {
	Fromname string `json:"fromName"`
	Toname   string `json:"toName"`
	Message  string `json:"message"`
}
```

Now we need to update the `processEvent()` function found on line 93 of the `main.go` file.

```go
// processEvent processes the event and returns the result of the processing.
func processEvent(e cloudevents.Event) (interface{} /*result*/, error) {
	tBegin := time.Now()
	// Create our new struct to hold the event data.
	ed := &EventData{}
	// Unmarshal the event data into our struct.
	if err := e.DataAs(ed); err != nil {
		return nil, err
	}

	// Do something with the event data.
	log.Printf("Received event with message of : %+v\n", ed.Message)
	newMessage := "Hello " + ed.Fromname + "! Thank you for the message!"

	res := &exampleResult{
		Message:        newMessage,
		ProcessingTime: time.Since(tBegin).Milliseconds(),
	}
	return res, nil
}
```

Remove the `randomDelay()` function as it is no longer needed.

> Run the application locally:

Author: I updated it but I like the demonstration of testing locally being there as well. what do you think of this? 
https://github.com/knative/docs/pull/4208/commits/04c094601726b4d7bd31b25ff0bdcfda23b7f722

> Run the application locally:

Reviewer: I'm with you on that. Although for consistency, it should be up around L180 as well as L298.

> Run the application locally:

Author: im with you on that... how about it all goes in one place? https://github.com/knative/docs/pull/4208/commits/63fbc06568d2be1b934a2f7bfde3f636fb33a62f

---
title: "Writing an event sink using Golang"
weight: 10
type: "docs"
showlandingtoc: "false"
---

> 

Reviewer 2: this frontmatter isn't required anymore since me moved from hugo to mkdocs

Author: fixed: https://github.com/knative/docs/pull/4208/commits/b6cadbe876d44a25e58478fb8732f39eee14f712
Thank you!

# Writing an event sink using Golang

> This tutorial provides instructions to build an event sink in Golang and implement it as a ksvc.

Reviewer 2: please write "Knative Service" rather than "ksvc" to avoid confusion

Author: Fixed: https://github.com/knative/docs/pull/4208/commits/b8225598113892d3f55f56b0d5d006485761e575 

Thank you!

> ## A Quick Overview of Building and Deploying a Simple Event Sink with Golang

Reviewer 2: Use sentence case for all headings:
"A quick overview of building and deploying a simple event sink with Golang"

Author: Fixed: https://github.com/knative/docs/pull/4208/commits/b8225598113892d3f55f56b0d5d006485761e575 
Thank you!

> Below we will create a simple Sink that receives CloudEvents and performs an example operation on them.

Reviewer 2: ```suggestion
In the following procedure, you can create a simple sink that receives CloudEvents and performs an example operation on them.
```

> From within the `$GOPATH`, Create a folder `demosink` with a new file `main.go`:

Reviewer 2: Make these numbered steps please so that it's easier for folks to follow

Author: Fixed: https://github.com/knative/docs/pull/4208/commits/6f43db24c1ef811a54d9de91acf0bcc1f2c9aa68 
Thank you!

> From within the `$GOPATH`, Create a folder `demosink` with a new file `main.go`:

Reviewer 2: ```suggestion
1. From within the `$GOPATH`, create a folder named `demosink`, that contains a new file named `main.go`:
```

	tBegin := time.Now()

	// simulate processing time
	randomDelay()

	res := &exampleResult{
		Message:        "event processed successfully",
		ProcessingTime: time.Since(tBegin).Milliseconds(),
	}

	return res, nil
}

// exampleResult represents a fictional structured result of some event
// processing.
type exampleResult struct {
	Message        string `json:"message"`
	ProcessingTime int64  `json:"processing_time_ms"`
}

// newAckEvent returns a CloudEvent that acknowledges the processing of an
// event.
func newAckEvent(e cloudevents.Event, code ACKCode, detail interface{}) *cloudevents.Event {
	resp := cloudevents.NewEvent()
	resp.SetType(eventTypeAck)
	resp.SetSource(eventSrcName)
	resp.SetExtension(ceExtProcessedType, e.Type())

	data := &ACKResponse{
		Code:   code,
		Detail: detail,
	}

	if err := resp.SetData(cloudevents.ApplicationJSON, data); err != nil {
		log.Panicf("Error serializing CloudEvent data: %s", err)
	}

	return &resp
}

// randomDelay sleeps for a random amount of time.
func randomDelay() {
	const maxDelay = 100 * time.Millisecond

	randomDelay := time.Duration(rand.Int63n(int64(maxDelay)))
	time.Sleep(randomDelay)
}

```

> Create the `go.mod` file and run `tidy`:

Reviewer 2: ```suggestion
1. Create the `go.mod` file, then run the following commands:
```
If you do add numbered steps to the doc, please indent the code samples so that the numbering renders properly also.

Author: Okay! Thank you!

```
go mod init
go mod tidy
```

> Create a new file `Dockerfile` in the same folder:
> ```dockerfile
> FROM golang:1.15-buster AS builder
> ENV CGO_ENABLED 0
> WORKDIR /project
> COPY . ./
> RUN go build
> FROM gcr.io/distroless/static:nonroot
> COPY --from=builder /project/demosink /
> ENTRYPOINT ["/demosink"]
> ```

Reviewer 2: ```suggestion
1. Create a new file `Dockerfile` in the same folder:
    
    ```dockerfile
    FROM golang:1.15-buster AS builder
    ENV CGO_ENABLED 0
    WORKDIR /project
    COPY . ./
    RUN go build
    FROM gcr.io/distroless/static:nonroot
    COPY --from=builder /project/demosink /
    ENTRYPOINT ["/demosink"]
    ```
```

suggestion of what I meant wrt. numbering and indentation 🙂

Reviewer 2: repeat for all below steps...let me know if you have any trouble; I can just add a clean up PR after this merges if you do.

Run the application locally to debug:

```sh
go run .
```

Send it an event and verify the expected results:
```sh
curl -v "http://localhost:8080" \
       -X POST \
       -H "Ce-Id: 536808d3-88be-4077-9d7a-a3f162705f79" \
       -H "Ce-Specversion: 1.0" \
       -H "Ce-Type: io.demo.email.send" \
       -H "Ce-Source: dev.knative.samples/demo" \
       -H "Content-Type: application/json" \
       -d '{"fromName":"richard","toName":"bob","message":"hello"}'
* upload completely sent off: 55 out of 55 bytes
< HTTP/1.1 200 OK
< Ce-Id: 70eb24d9-3678-4c08-9332-315cbae7fe1e
< Ce-Processedid: 536808d3-88be-4077-9d7a-a3f162705f79
< Ce-Processedsource: dev.knative.samples/demo
< Ce-Processedtype: io.demo.email.send
< Ce-Source: io.demo.targets.go-sample
< Ce-Specversion: 1.0
< Ce-Time: 2021-09-08T19:40:37.1619Z
< Ce-Type: com.example.target.ack
< Content-Length: 98
< Content-Type: application/json
< Date: Wed, 08 Sep 2021 19:40:37 GMT
<
* Connection #0 to host localhost left intact
{"code":0,"detail":{"message":"Hello richard! Thank you for the message!","processing_time_ms":0}}* Closing connection 0
```

When you are ready to deploy, build and publish the container image:

Google:
```sh

> gcloud builds submit --tag gcr.io/<project-name>/<image-name> .

Reviewer 2: All of our documentation should be vendor neutral, so if this sample requires Google platform specific tools etc please modify it so that there is a generic version instead, e.g. using `kubectl` commands.

Author: Fixed by removing the `gcloud` reference: https://github.com/knative/docs/pull/4208/commits/891e4c9ba7c5f72755deebdae93b81a2c7866c73 
Thank you!
